File: backend/app/api/v1/schemas/screener.py
"""Screener request/response models."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class ScreenerQuery(BaseModel):
    symbol: str = Field(min_length=1, max_length=10)
    min_strike: float | None = Field(default=None, gt=0)
    max_strike: float | None = Field(default=None, gt=0)
    expiry_days_min: int = Field(default=0, ge=0, le=365)
    expiry_days_max: int = Field(default=60, ge=0, le=730)
    option_type: Literal["call", "put"] | None = None
    min_volume: int = Field(default=0, ge=0)
    min_open_interest: int = Field(default=0, ge=0)
    max_iv: float | None = Field(default=None, gt=0, le=5)
    min_iv: float | None = Field(default=None, gt=0, le=5)
    limit: int = Field(default=50, ge=1, le=500)


class ScreenerResultRow(BaseModel):
    symbol: str
    strike: float
    expiry: str
    option_type: Literal["call", "put"]
    bid: float
    ask: float
    last_price: float
    implied_vol: float
    delta: float
    gamma: float
    theta: float
    vega: float
    open_interest: int
    volume: int


class ScreenerResponse(BaseModel):
    success: bool = True
    symbol: str
    spot: float
    count: int
    results: list[ScreenerResultRow]

File: frontend/src/types/screener.ts
/** Mirrors `backend/app/api/v1/schemas/screener.py`. */

import type { OptionType } from "./common";

export interface ScreenerFilters {
  symbol: string;
  min_strike?: number | null;
  max_strike?: number | null;
  expiry_days_min: number;
  expiry_days_max: number;
  option_type?: OptionType | null;
  min_volume: number;
  min_open_interest: number;
  max_iv?: number | null;
  min_iv?: number | null;
  limit: number;
}

export interface ScreenerRow {
  symbol: string;
  strike: number;
  expiry: string;
  option_type: OptionType;
  bid: number;
  ask: number;
  last_price: number;
  implied_vol: number;
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  open_interest: number;
  volume: number;
}

export interface ScreenerResponse {
  success: true;
  symbol: string;
  spot: number;
  count: number;
  results: ScreenerRow[];
}
